### PR TITLE
enh(markdown/xml) auto-detection fixes

### LINKED ISSUE #2767: (xml) Grammar is far too broad and detects far to many things
**Describe the issue**

Via StackOverflow https://meta.stackexchange.com/a/355120/853855

This code is auto-detected as markdown (with embedded XML):

```
typedef float pnt[3];

void count(pnt const * pnts, const int n, unsigned cnt[8]) {
 
  for(int i=0; i<8; ++i)
    cnt[i] = 0;

  for(int i=0; i<n; ++i)
    if(pnts[i][0]>=0.0f && pnts[i][1]>=0.0f && pnts[i][2]>=0.0f ) ++cnt[7]; else
    if(pnts[i][0]>=0.0f && pnts[i][1]>=0.0f && pnts[i][2]< 0.0f ) ++cnt[3]; else
    if(pnts[i][0]>=0.0f && pnts[i][1]< 0.0f && pnts[i][2]>=0.0f ) ++cnt[5]; else
    if(pnts[i][0]>=0.0f && pnts[i][1]< 0.0f && pnts[i][2]< 0.0f ) ++cnt[1]; else
    if(pnts[i][0]< 0.0f && pnts[i][1]>=0.0f && pnts[i][2]>=0.0f ) ++cnt[6]; else
    if(pnts[i][0]< 0.0f && pnts[i][1]>=0.0f && pnts[i][2]< 0.0f ) ++cnt[2]; else
    if(pnts[i][0]< 0.0f && pnts[i][1]< 0.0f && pnts[i][2]>=0.0f ) ++cnt[4]; else
    ++cnt[0];
}
```

**Which language seems to have the issue?**

The core problem is that XML's rules are far too broad and match things that have no chance in heck of being XML.

**Are you using `highlight` or `highlightAuto`?**

This is an auto-detect issue with `xml` being far to greedy.  The "name" if tags aren't even checked to see if they are valid... ie in the above example `<8;>` is not a valid HTML tag (the `;` should be a dead giveaway)...

**Expected behavior**

XML ruleset should be tightened.

### PR DESCRIPTION
**Pull request by joshgoebel**
Markdown is far too greedy (relevance) and XML is far too loose about what it considers XML.  Together they make a dangerous combo. 

Resolves #2767.

### Changes

- improve precision of XML to prevent false auto-detect positives (much tighter ruleset)
- greatly reduce runaway relevancy of Markdown (+200 to 0 on the test snippet, `cpp` [the correct language] scores a 17)
- links should not have 10 relevancy
- `[x][i]` pattern reduced to 0 relevancy, this can just be too many things in so many languages (array access, etc)
- `[x][commonscheme://]` pattern gets 2 relevancy (data, http, https, ftp, etc)
- `[x][neverhardof://]` pattern gets 2 relevancy (ie `.*(://.*)`), stills looks enough like a URL, etc.
- `[x][relative url]` pattern gets 1 relevancy
- `[x](x)` pattern gets default 1 relevancy

### Checklist
- [x] Added markup tests, or they don't apply here because output remains unchanged
- [x] Updated the changelog at `CHANGES.md`
- [x] Added myself to `AUTHORS.txt`, under Contributors

### CODE REVIEW

**Comment by joshgoebel** [comment]
Context:

File: src/languages/markdown.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ Category: common, markup
 
 export default function(hljs) {
   const INLINE_HTML = {
-    begin: '<', end: '>',
+    begin: '<[A-Za-z_]', end: '>',
```
Comment (joshgoebel) [inline src/languages/markdown.js:11]: Should reduce a lot of false positives on `<` anything else.

File: src/languages/markdown.js [hunk 2]
```diff
@@ -71,16 +76,19 @@ export default function(hljs) {
       },
       {
         className: 'link',
+        relevance: 0,
```
Comment (joshgoebel) [inline src/languages/markdown.js:79]: The relevance here all happens in the top LINK rule, not in the children.  Relevance here is counted a single time, not layered.

**Comment by joshgoebel** [comment]
Ping.

**Comment by taufik-nurrohman** [comment]
> If we're really worried we could add a `[x](http:...)` pattern or similar.

~~~ js
\((?:data|(?:http|ftp)s?|javascript|mailto):\/\/.*?\)
~~~

**Comment by joshgoebel** [comment]
Are you suggesting those to boost relevance and filter false positives?

**Comment by joshgoebel** [comment]
@taufik-nurrohman Better?

**Comment by joshgoebel** [comment]
@egor-rogov @allejo Going to merge this based on @taufik-nurrohman review if you all don't have any time in the next few days to have a look.  Thanks!

**Comment by taufik-nurrohman** [comment]
Sorry. There is something wrong about the protocol pattern: `data:` and `javascript:` aren&rsquo;t followed by `//`

**Comment by joshgoebel** [comment]
I don't see the problem, looks and works properly here...

**Comment by taufik-nurrohman** [comment]
~~~ .js
​\(​(?:(​?:​http​|​ftp​)​s​?:\/\/​|(?:data|javascript|mailto):).​*?\)
~~~

**Comment by joshgoebel** [comment]
I like the //... so:

```
\[.+?\]\(((data|javascript|mailto):|(?:http|ftp)s?:\/\/).*?\)
```

Look good?

**Comment by joshgoebel** [comment]
@allejo Any chance you have time to review?

**Comment by allejo** [comment]
Sure thing! Sorry, been trying to catch up with review requests here. I'll do this one next

File: src/languages/xml.js [hunk 3]
```diff
@@ -124,16 +129,48 @@ export default function(hljs) {
...
+          starts: TAG_INTERNALS
+        }]
+      },
+      // close tag
+      {
+        className: 'tag',
+        begin: regex.concat(/<\//,
+          regex.lookahead(regex.concat(
+            TAG_NAME_RE, />/)
+          )),
```
Comment (allejo) [inline src/languages/xml.js:162]:
```suggestion
        begin: regex.concat(
          /<\//,
          regex.lookahead(regex.concat(
            TAG_NAME_RE, />/
          ))
        ),
```

File: src/languages/xml.js [hunk 4]
```diff
@@ -124,16 +129,48 @@ export default function(hljs) {
...
+      {
+        className: 'tag',
+        begin: regex.concat(/</,
+          regex.lookahead(regex.concat(
+            TAG_NAME_RE,
+            // <tag/>
+            // <tag>
+            // <tag ...
+            regex.either(/\/>/, />/, /\s/))
+          )),
```
Comment (allejo) [inline src/languages/xml.js:147]:
```suggestion
        begin: regex.concat(
          /</,
          regex.lookahead(regex.concat(
            TAG_NAME_RE,
            // <tag/>
            // <tag>
            // <tag ...
            regex.either(/\/>/, />/, /\s/)
          ))
        ),
```

File: src/languages/xml.js [hunk 5]
```diff
@@ -124,16 +129,48 @@ export default function(hljs) {
...
+        contains: [{
+          className: 'name',
+          begin: TAG_NAME_RE,
+          relevance: 0
+        },
+        {
+          begin: />/,
+          relevance: 0
+        }
+      ]
```
Comment (allejo) [inline src/languages/xml.js:172]:
```suggestion
        contains: [
          {
            className: 'name',
            begin: TAG_NAME_RE,
            relevance: 0
          },
          {
            begin: />/,
            relevance: 0
          }
        ]
```

**Review by allejo — APPROVED** [review]
Aside from code formatting that confused me while reading this on GH, this looks good to me

**Comment by joshgoebel** [comment]
@allejo Do I have to do something special to make the new actions fire here?

**Pull request merged**